crates/rome_js_analyze/src/analyzers/nursery/no_fallthrough_switch_clause.rs

```diff
@@ -88,31 +206,82 @@ impl Rule for NoFallthroughSwitchClause {
     }
 }
 
-fn case_fell(case: &AnyJsSwitchClause) -> bool {
-    let statements = case.consequent();
-    !has_fall_blocker_statement(&statements) && statements.iter().count() != 0
-}
-
-fn has_fall_blocker_statement(statements: &JsStatementList) -> bool {
-    for statement in statements.iter() {
-        if is_fall_blocker_statement(&statement) {
-            return true;
-        }
-        if let Some(block_statement) = JsBlockStatement::cast_ref(statement.syntax()) {
-            if has_fall_blocker_statement(&block_statement.statements()) {
+fn has_switch_statement(control_flow_root: &JsSyntaxNode) -> bool {
+    let mut iter = control_flow_root.preorder();
+    while let Some(node) = iter.next() {
+        if let WalkEvent::Enter(node) = node {
+            if JsSwitchStatement::can_cast(node.kind()) {
                 return true;
+            } else if AnyJsControlFlowRoot::can_cast(node.kind()) && control_flow_root != &node {
+                iter.skip_subtree()
             }
         }
     }
     false
 }
 
-fn is_fall_blocker_statement(statement: &AnyJsStatement) -> bool {
-    matches!(
-        statement,
-        AnyJsStatement::JsBreakStatement(_)
-            | AnyJsStatement::JsReturnStatement(_)
-            | AnyJsStatement::JsThrowStatement(_)
-            | AnyJsStatement::JsContinueStatement(_)
-    )
+fn register_fallthrough_switch_clauses(
+    block_to_switch_clause_range: &FxHashMap<u32, Option<TextRange>>,
+    visited_blocks: &FxHashSet<u32>,
+    cfg: &rome_control_flow::ControlFlowGraph<JsLanguage>,
+    fallthrough: &mut Vec<TextRange>,
+) {
+    let mut current_switch_clause = None;
+    // Register all switch clauses as block to process.
+    let mut block_stack: Vec<u32> = block_to_switch_clause_range.keys().copied().collect();
+    let mut visited_blocks = visited_blocks.clone();
+    // Traverse the control flow graph
+    while let Some(block_index) = block_stack.pop() {
+        current_switch_clause = block_to_switch_clause_range
+            .get(&block_index)
+            .or(current_switch_clause);
+        // SAFETY: this is a safe conversion because it is already an index for `cfg.blocks`.
+        let block_index = block_index as usize;
+        let Some(block) = cfg.blocks.get(block_index) else {
+            continue;
+        };
+        // Register exception handlers as blocks to process
+        // Ignore finally handler: they are already in the Control Flow Graph.
+        for exception_handler in block
+            .exception_handlers
+            .iter()
+            .filter(|x| matches!(x.kind, ExceptionHandlerKind::Catch))
+        {
+            // Avoid cycles and redundant checks.
+            if visited_blocks.insert(exception_handler.target) {
+                block_stack.push(exception_handler.target);
+            }
+        }
```

Comment:
There's some repeated code here. It seems that we are going the same operation, twice. Why is that? You should assume that I have little knowledge of this new logic.

Reply:
The first traversal detects switch and the corresponding blocks of its switch clauses. This traversal detects fallthroughs.